.github/workflows/main.yml

```diff
@@ -1,65 +1,62 @@
-# copy of https://github.com/actions-rs/meta/blob/master/recipes/quickstart.md
 on: [push, pull_request]
 
 name: Continuous integration
 
 jobs:
-  check:
-    name: Check
-    strategy:
-      matrix:
-        os: [ubuntu-latest, windows-latest, macos-latest]
-    runs-on: ${{ matrix.os }}
-    steps:
-      - uses: actions/checkout@v1
-      - uses: actions-rs/toolchain@v1
-        with:
-          profile: minimal
-          toolchain: stable
-          override: true
-      - uses: actions-rs/cargo@v1
-        continue-on-error: false
-        with:
-          command: check
-
   test:
-    name: Test Suite
+    name: Test
     strategy:
       matrix:
         os: [ubuntu-latest, windows-latest, macos-latest]
     runs-on: ${{ matrix.os }}
     steps:
-      - uses: actions/checkout@v1
-      - uses: actions-rs/toolchain@v1
-        with:
-          profile: minimal
-          toolchain: stable
-          override: true
-      - uses: actions-rs/cargo@v1
-        continue-on-error: false
+
+      - uses: actions/checkout@v2
+
+      - name: Fetch commit hash of the crates.io registry index
+        shell: bash
+        run: |
+          commit=$(git ls-remote --heads https://github.com/rust-lang/crates.io-index.git master | cut -f1)
+          echo "$commit"
+          echo "::set-env name=CRATES_IO_SHA::$commit"
+
+      - name: Cache cargo registry
+        uses: actions/cache@v1
         with:
-          command: test
+          path: ~/.cargo/registry
+          key: ${{ runner.os }}-cargo-registry-v1-${{ hashFiles('**/Cargo.toml') }}-${{ env.CRATES_IO_SHA }}
+          restore-keys: |
+            ${{ runner.os }}-cargo-registry-v1-${{ hashFiles('**/Cargo.toml') }}-
+            ${{ runner.os }}-cargo-registry-v1-
 
-  lints:
-    name: Rust lints
-    runs-on: ubuntu-latest
-    steps:
-      - uses: actions/checkout@v1
-      - uses: actions-rs/toolchain@v1
+      - name: Install Rust toolchain
+        uses: actions-rs/toolchain@v1
         with:
           profile: minimal
           toolchain: stable
           override: true
           components: rustfmt, clippy
 
-      - name: Run cargo fmt
+      - name: Check code formatting
         uses: actions-rs/cargo@v1
         continue-on-error: false
         with:
           command: fmt
           args: --all -- --check
 
-      - name: Run cargo clippy
+      - name: Check build
+        uses: actions-rs/cargo@v1
+        continue-on-error: false
+        with:
+          command: check
+
+      - name: Run test suite
+        uses: actions-rs/cargo@v1
+        continue-on-error: false
+        with:
+          command: test
+
+      - name: Check clippy lints
```

Comment:
I think that by using jobs if the first one fails the others will not even be run. So maybe we can move the lints to the top, or after `cargo fmt` so it is not blocked by the build/test

Reply:
Maybe, but now it only dumps warnings into GitHub logs. When we are ready to deny clippy warnings, we can move it to the top after the `fmt` check.

Reply:
Ah yes its true, it just dump warnings.